torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -10,167 +10,609 @@
+  }
+  auto& state = static_cast<LBFGSParamState&>(*state_[c10::guts::to_string(_params.at(0).unsafeGetTensorImpl())]);
+  // evaluate initial f(x) and df/dx
+  Tensor orig_loss;
+  {
+    torch::AutoGradMode enable_grad(true);
+    orig_loss = closure();
   }
 
+  auto loss = val(orig_loss);
+  auto current_evals = 1;
+  state.func_evals(state.func_evals() + 1);
+  auto flat_grad = _gather_flat_grad();
+  auto opt_cond = (val(flat_grad.abs().max()) <= tolerance_grad);
+
+  // optimal condition
+  if (opt_cond) {
+    return orig_loss;
+  }
+
+  // tensors cached in state (for tracing)
+  auto& d = state.d();
+  auto& t = state.t();
+  auto& old_dirs = state.old_dirs();
+  auto& old_stps = state.old_stps();
+  auto& ro = state.ro();
+  auto& H_diag = state.H_diag();
+  auto& prev_flat_grad = state.prev_flat_grad();
+  auto& prev_loss = state.prev_loss();
+
+  int n_iter = 0;
   Tensor ONE = torch::tensor(1, flat_grad.options());
 
-  int64_t n_iter = 0;
-  while (n_iter < options.max_iter()) {
-    n_iter++;
-    state_n_iter++;
+  // optimize for a max of max_iter iterations
+  while(n_iter < max_iter) {
+    // keep track of nb of iterations
+    n_iter += 1;
+    state.n_iter(state.n_iter() + 1);
 
-    if (state_n_iter == 1) {
+    // compute gradient descent direction
+    if (state.n_iter() == 1) {
       d = flat_grad.neg();
       H_diag = ONE;
-      prev_flat_grad = flat_grad.clone(at::MemoryFormat::Contiguous);
     } else {
-      Tensor y = flat_grad.sub(prev_flat_grad);
-      Tensor s = d.mul(t);
-      Tensor ys = y.dot(s);
-
-      if (ys.item<float>() > 1e-10) {
+      // do lbfgs update (update memory)
+      auto y = flat_grad.sub(prev_flat_grad);
+      auto s = d.mul(t);
+      auto ys = y.dot(s); // y*s
+      if (val(ys) > 1e-10) { // NOLINT(cppcoreguidelines-avoid-magic-numbers)
         // updating memory
-
-        if (old_dirs.size() == options.history_size()) {
-          // shift history by one (limited memory)
+        if (int(old_dirs.size()) == history_size) {
+          // shift history by one (limited-memory)
           old_dirs.pop_front();
           old_stps.pop_front();
+          ro.pop_front();
         }
-
         // store new direction/step
-        old_dirs.push_back(y);
-        old_stps.push_back(s);
+        old_dirs.emplace_back(y);
+        old_stps.emplace_back(s);
+        ro.emplace_back(1. / ys);
 
         // update scale of initial Hessian approximation
-        H_diag = ys / y.dot(y);
+        H_diag = ys / y.dot(y);  // (y*y)
       }
 
-      int64_t num_old = old_dirs.size();
+      // compute the approximate (L-BFGS) inverse Hessian
+      // multiplied by the gradient
+      auto num_old = old_dirs.size();
 
-      for (int64_t i = 0; i < num_old; i++) {
-        ro.at(i) = ONE / old_dirs.at(i).dot(old_stps.at(i));
+      auto& al = state.al();
+      if (al != c10::nullopt) {
+        (*al).resize(history_size);
       }
```

Comment:
Should this be the following instead?
```suggestion
      auto& al = state.al();
      if (al == c10::nullopt) {
        al = std::vector<Tensor>(history_size, Tensor());
      }
```
otherwise I don't think `al`'s value would ever change from `c10::nullopt` to some `std::vector<Tensor>` :D